pkg/data/generic.go

```diff
@@ -79,167 +87,175 @@ func (s *GenericSync) RunContext(ctx context.Context) error {
 	if s.createError != nil {
 		return s.createError
 	}
-	s.loop(ctx.Done())
+
+	store, queue := s.setup(ctx)
+	defer queue.ShutDown()
+
+	go s.loop(ctx, store, queue)
+	<-ctx.Done()
```

Comment:
🤔 At this point, we could also just have loop return, I think?

```go
s.loop(ctx, store, queue)
return nil
```
or even 
```go
return s.loop(ctx, store.queue)
```
?

But there may be other users of `s.loop` so we can't do this? 🤔

Reply:
We wait for `<-ctx.Done()` before returning because of the `defer queue.ShutDown()`. Cancelling the context is not enough to stop `loop`, we also needed to wait for the context to end and then shutdown the queue.

I have made it more explicit in c5ef3a752ba1fd2f2fee267ec2902267396ca21c. Removed the unused `ctx` parameter from loop, and moved the queue shutdown to a separate goroutine.